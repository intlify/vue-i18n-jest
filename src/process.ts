import { readFileSync } from 'fs'
import { isAbsolute as isAbsolutePath, resolve as resolvePath } from 'path'

import { debug as Debug } from 'debug'
import { parse as parseJSON5 } from 'json5'
import { safeLoad as parseYAML } from 'js-yaml'
import type { SFCBlock } from 'vue-template-compiler'

const debug = Debug('vue-i18n-jest')

type SFCCustomBlock = Pick<SFCBlock, 'attrs' | 'content' | 'type'>

type ProcessParameter = {
  /** All of the blocks matching your type, returned from `@vue/component-compiler-utils` */
  blocks: SFCCustomBlock[]
  /** The internal namespace for a component's Vue Options in vue-jest */
  vueOptionsNamespace: string
  /** The SFC file being processed */
  filename: string
}

const VUE_I18N_OPTION = '__i18n'

/**
 * Process vue-i18n contents inside of a custom block and prepare it for execution in a testing environment.
 */
export default function process ({ blocks, vueOptionsNamespace, filename }: ProcessParameter) : string {
  const i18nResources = blocks.map(block => {
    if (block.type !== 'i18n') return

    const value = parseI18nBlockToJSON(block, filename)
      .replace(/\u2028/g, '\\u2028') // LINE SEPARATOR
      .replace(/\u2029/g, '\\u2029') // PARAGRAPH SEPARATOR
      .replace(/\\/g, '\\\\')
      .replace(/'/g, "\\'")
    return `'${value}'`
  }).filter((s): s is string => !!s)

  // vueOptions.__i18n = [
  // '<json encoded block 1>',
  // '<json encoded block 2>'
  // ]
  const i18nOption = `${vueOptionsNamespace}.${VUE_I18N_OPTION}`
  const code = i18nResources.length ? `${i18nOption} = [\n${i18nResources.join(',\n')}\n]` : ''
  debug('generatedCode', code)
  return code
}

/**
 * Parse custom `<i18n>` block content to JSON string.
 * @param block SFC block returned from `@vue/component-compiler-utils`
 * @param filename The SFC file being processed
 */
function parseI18nBlockToJSON (block: SFCCustomBlock, filename: string): string {
  const lang = block.attrs && block.attrs.lang
  const src = block.attrs && block.attrs.src
  const content = src
    ? readFileSync(getAbsolutePath(src, filename)).toString()
    : block.content

  return convertToJSON(content, lang)
}

/**
 * Convert JSON/YAML/JSON5 to minified JSON string.
 * @param source JSON/YAML/JSON5 encoded string
 * @param lang Language used in `source`. Supported JSON, YAML or JSON5.
 * @returns {string} A minified JSON string
 */
function convertToJSON (source: string, lang: string): string {
  switch (lang) {
    case 'yaml':
    case 'yml':
      return JSON.stringify(parseYAML(source))
    case 'json5':
      return JSON.stringify(parseJSON5(source))
    default: // fallback to 'json'
      return JSON.stringify(JSON.parse(source))
  }
}

function getAbsolutePath (src: string, fileName: string): string {
  return isAbsolutePath(src) ? src : resolvePath(fileName, '../', src)
}
