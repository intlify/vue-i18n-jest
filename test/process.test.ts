import { readFileSync } from 'fs'
import { resolve as resolvePath } from 'path'
import { parseComponent } from 'vue-template-compiler'

import process from '../src/process'

const files = [
  'no-i18n.vue',
  'default.vue',
  'json.vue',
  'json-src-import.vue',
  'json-locale.vue',
  'yaml.vue',
  'yaml-src-import.vue',
  'yaml-locale.vue',
  'json5.vue',
  'json5-src-import.vue',
  'json5-locale.vue'
]

describe('process', () => {
  test.each(files)('(%s) returns expected string', file => {
    const filePath = resolvePath(__dirname, `./fixtures/${file}`)
    const fileContent = readFileSync(filePath).toString()
    const { customBlocks: blocks } = parseComponent(fileContent)

    const code = process({ blocks, componentNamespace: 'Component', filename: filePath })

    expect(code).toMatchSnapshot(file)
  })
})
