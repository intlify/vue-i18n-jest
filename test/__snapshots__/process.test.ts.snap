// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process (default.vue) returns expected string: default.vue 1`] = `
"__vue__options__.__i18n = [
'{\\"ja\\":{\\"hello\\":\\"こんにちは！\\"},\\"en\\":{\\"hello\\":\\"hello!\\"}}'
]"
`;

exports[`process (json.vue) returns expected string: json.vue 1`] = `
"__vue__options__.__i18n = [
'{\\"ja\\":{\\"hello\\":\\"こんにちは！\\"},\\"en\\":{\\"hello\\":\\"hello!\\"}}'
]"
`;

exports[`process (json-src-import.vue) returns expected string: json-src-import.vue 1`] = `
"__vue__options__.__i18n = [
'{\\"ja\\":{\\"hello\\":\\"こんにちは！\\"},\\"en\\":{\\"hello\\":\\"hello!\\"}}'
]"
`;

exports[`process (json5.vue) returns expected string: json5.vue 1`] = `
"__vue__options__.__i18n = [
'{\\"ja\\":{\\"hello\\":\\"こんにちは！\\"},\\"en\\":{\\"hello\\":\\"hello!\\"}}'
]"
`;

exports[`process (json5-src-import.vue) returns expected string: json5-src-import.vue 1`] = `
"__vue__options__.__i18n = [
'{\\"ja\\":{\\"hello\\":\\"こんにちは！\\"},\\"en\\":{\\"hello\\":\\"hello!\\"}}'
]"
`;

exports[`process (no-i18n.vue) returns expected string: no-i18n.vue 1`] = `""`;

exports[`process (yaml.vue) returns expected string: yaml.vue 1`] = `
"__vue__options__.__i18n = [
'{\\"ja\\":{\\"hello\\":\\"こんにちは！\\"},\\"en\\":{\\"hello\\":\\"hello!\\"}}'
]"
`;

exports[`process (yaml-src-import.vue) returns expected string: yaml-src-import.vue 1`] = `
"__vue__options__.__i18n = [
'{\\"ja\\":{\\"hello\\":\\"こんにちは！\\"},\\"en\\":{\\"hello\\":\\"hello!\\"}}'
]"
`;
